File: front/src/app/services/absence.service.ts
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable, catchError, throwError } from 'rxjs';

import { environment } from '../../environments/environment';
import { Absence, AbsenceDetail, CreateAbsencePayload, MarkAbsencePayload, StudentStats } from '../models/absence.model';

@Injectable({ providedIn: 'root' })
export class AbsenceService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiUrl.replace(/\/$/, '')}/absences`;

  createAbsence(payload: CreateAbsencePayload): Observable<Absence> {
    return this.http.post<Absence>(this.baseUrl, payload).pipe(catchError(this.handleError));
  }

  markAbsences(absenceId: number, payload: MarkAbsencePayload): Observable<Absence> {
    return this.http.put<Absence>(`${this.baseUrl}/${absenceId}/mark`, payload).pipe(catchError(this.handleError));
  }

  getAbsencesByTeacher(teacherId: number): Observable<Absence[]> {
    return this.http.get<Absence[]>(`${this.baseUrl}/teacher/${teacherId}`).pipe(catchError(this.handleError));
  }

  getAbsencesByFiliereAndPromo(filiere: string, promo: string): Observable<Absence[]> {
    return this.http.get<Absence[]>(`${this.baseUrl}/filiere/${encodeURIComponent(filiere)}/promo/${encodeURIComponent(promo)}`).pipe(catchError(this.handleError));
  }

  getAbsenceDetails(absenceId: number): Observable<AbsenceDetail[]> {
    return this.http.get<AbsenceDetail[]>(`${this.baseUrl}/${absenceId}/details`).pipe(catchError(this.handleError));
  }

  getStudentStats(studentId: number): Observable<StudentStats> {
    return this.http.get<StudentStats>(`${this.baseUrl}/student/${studentId}/stats`).pipe(catchError(this.handleError));
  }

  private handleError(error: HttpErrorResponse) {
    const message = error.error?.message || error.error?.error || 'Unable to process attendance request.';
    return throwError(() => new Error(message));
  }
}

File: backend/src/main/java/org/example/universityattendancemanagementsystem/service/AbsenceService.java
package org.example.universityattendancemanagementsystem.service;

import org.example.universityattendancemanagementsystem.dto.AbsenceDetailDto;
import org.example.universityattendancemanagementsystem.dto.AbsenceDto;

import java.time.LocalDate;
import java.util.List;
import java.util.Map;

public interface AbsenceService {
    AbsenceDto createAbsence(Long teacherId, String nomModule, String filiere, String promo, String typeSeance, LocalDate date, List<Long> studentIds);

    AbsenceDto markAbsences(Long absenceId, List<Long> absentStudentIds);

    List<AbsenceDto> getAbsencesByTeacher(Long teacherId);

    List<AbsenceDto> getAbsencesByFiliereAndPromo(String filiere, String promo);

    List<AbsenceDetailDto> getAbsenceDetails(Long absenceId);

    Map<String, Long> countAbsencesByStudent(Long studentId);
}
